content/articles/web-apps.md

```diff
@@ -0,0 +1,108 @@
+
+However, in cases where a simple application is all that is required, with no complex interactivity or scalability requirements, a Shiny application may be the best option for the final application too.
+
+# Components of a full research web application
+
+Our web application installations typically consist of three main components:
+- research code - provided by the research team
+- an interface layer - this provides a way for the web application to provide inputs and receive outputs from the research code. It is usually written in the same language as the research code and invokes it directly, or via a task queue. The interface layer usually provides a local HTTP API which is accessible to the web application
+- web application, consisting of two parts:
+    - back end: e.g kotlin or express.js, which forwards requests to the interface layer, and can do various bits of housekeeping, like managing persistent user data or uploaded files.
+    - front end: e.g Vue.js or React which is loaded in the user's browser and provides the user interface and visualisations.
+
+We run all of these components within docker containers to provide reliable portability between installation environments. Typically the front and back ends of the web applications will go into one container, and the research code and interface layer will go into another. We'll run a web proxy in a third container, and there may be additional containers for databases, task queues etc.
+
+
+# Pre-requisites
+
+In order to support this architecture, we require the research code to have the following features so that we can build a stable interface to it:
+
+- Source-controlled in github: we maintain the source code for our applications in github repositories, and we require research code to be available in github too, so that we can develop against it, apply automated tests to it within continuous integration and deploy it using docker.
+
+- Versioned: it's useful to be able surface a version number in the application, of both the research and application code. It's helpful if the research code can provide its own semantic version, but as a minimum we could use a github commit identifier.
+
+- Data sets are available and clearly defined: any data required for the application (data which is used by the research code, or has been generated by it) should be available to automated access for CI and deployment. This could be on a network drive, but it would be preferable if it was in something like a github release. Data should also be versionsed in some way, and its format should be identical between versions, except for known format upgrades.
```

Comment:
```suggestion
- Data sets are available and clearly defined: any data required for the application (data which is used by the research code, or has been generated by it) should be available to automated access for CI and deployment. This could be on a network drive, but it would be preferable if it was in something like a github release. Data should also be versioned in some way, and its format should be identical between versions, except for known format upgrades.
```